core/src/test/java/org/apache/calcite/test/RelBuilderTest.java

```diff
@@ -4031,6 +4033,22 @@ private static RelNode groupIdRel(RelBuilder builder, boolean extra) {
     assertThat(root, hasTree(expected));
   }
 
+  @Test void testFilterInCharWithDiffLength() {
+    final Function<RelBuilder, RelNode> f = b ->
+        b.scan("EMP")
+            .filter(b.in(b.field("JOB"), b.literal("CLERK"), b.literal("A")))
+            .build();
+    final String expected =
+        "LogicalFilter(condition=[SEARCH($2, Sarg['A':CHAR(5), 'CLERK']:CHAR(5))])\n"
+            + "  LogicalTableScan(table=[[scott, EMP]])\n";
+    final String expectedWithLeastRestrictive =
+        "LogicalFilter(condition=[SEARCH($2, Sarg['A':VARCHAR(5), 'CLERK':VARCHAR(5)]:VARCHAR(5))])\n"
+        + "  LogicalTableScan(table=[[scott, EMP]])\n";
+    assertThat(f.apply(createBuilder()), hasTree(expected));
+    assertThat(f.apply(RelBuilder.create(config().typeSystem(VaryingTypeSystem.INSTANCE).build())),
+        hasTree(expectedWithLeastRestrictive));
+  }
+
```

Comment:
Can you add a case for `(a,b) in (('1','1'), ('22', '333'))`?

Reply:
@hsyuan It seems that the `RexNode` which represents a filter condition like `(a,b) in (('1','1'), ('22', '333'))` could not created by `RexBuilder.makeIn` because the condition would be translated into a composed condition like 
`OR(AND(=($0, _UTF-16LE'1'), =($1, _UTF-16LE'11')), AND(=($0, _UTF-16LE'2'), =($1, _UTF-16LE'333')))`. right?

Reply:
@beyond1920 Before the code review, I think we need to reach an agreement about the method that how to resolve this issue.

Reply:
@NobiGo Thanks for comment. I left some message in JIRA, welcome to further discussion.